node/pkg/zeropglog/app.go

```diff
@@ -0,0 +1,188 @@
+package zeropglog
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+	"strings"
+	"time"
+
+	"bisonai.com/orakl/node/pkg/db"
+	errorsentinel "bisonai.com/orakl/node/pkg/error"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+func New(options ...AppOption) *App {
+	c := &AppConfig{
+		StoreInterval: DefaultLogStoreInterval,
+		Buffer:        DefaultBufferSize,
+	}
+	for _, option := range options {
+		option(c)
+	}
+	return &App{
+		StoreInterval: c.StoreInterval,
+		buffer:        make(chan map[string]any, c.Buffer),
+	}
+}
+
+func (a *App) Write(p []byte) (n int, err error) {
+	res, err := byte2Entry(p)
+	if err != nil {
+		return 0, err
+	}
+	a.buffer <- res
+	return len(p), nil
+}
+
+func (a *App) Run(ctx context.Context) {
+	a.setup()
+
+	ticker := time.NewTicker(a.StoreInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			err := a.processBatch(ctx)
+			if err != nil {
+				log.Err(err).Msg("log batch process failure")
+			}
+
+		}
+	}
+}
+
+func (a *App) setup() {
+	zerolog.TimeFieldFormat = zerolog.TimeFormatUnix
+	a.setLogLevel()
+	a.setLogWriter()
+}
+
+func (a *App) setLogLevel() {
+	logLevel := os.Getenv("LOG_LEVEL")
+	if logLevel == "" {
+		logLevel = "info"
+	}
+	zerolog.SetGlobalLevel(getLogLevel(logLevel))
+}
+
+func (a *App) setLogWriter() {
+	consoleWriter := zerolog.ConsoleWriter{Out: os.Stdout}
+	multiWriter := zerolog.MultiLevelWriter(consoleWriter, a)
+	logger := zerolog.New(multiWriter).With().Timestamp().Logger()
+	log.Logger = logger
+}
+
+func (a *App) processBatch(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return nil
+	case entry := <-a.buffer:
+		batch := []map[string]any{entry}
+	loop:
+		for {
+			select {
+			case entry := <-a.buffer:
+				batch = append(batch, entry)
+			default:
+				break loop
+			}
+		}
+		return bulkCopyLogEntries(ctx, batch)
+	default:
+		return nil
+	}
+}
+
+func bulkCopyLogEntries(ctx context.Context, logEntries []map[string]any) error {
+	bulkCopyEntries := [][]any{}
+
+	for _, entry := range logEntries {
+		res, err := extractDbEntry(entry)
+		if err != nil {
+			log.Error().Err(err).Msg("Error extracting log entry")
+			continue
+		}
+
+		if res[1].(zerolog.Level) < zerolog.WarnLevel {
```

Comment:
I'm wondering, how about checking this at the beginning of the for loop before `extractDbEntry` is called? That could save unnecessary processing given that we have a lot of `debug` and `info` level logs

Reply:
agreed